core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/common/index/IndexerFactory.java

```diff
@@ -1,115 +1,568 @@
+ * Each {@link AbstractTuple tuple} is assigned an {@link IndexProperties} instance,
+ * which determines its location in the index.
+ * {@link IndexProperties} instances are built from {@link AbstractJoiner joiners}
+ * using methods such as {@link #buildUniLeftMapping()} and {@link #buildRightMapping()}.
+ * Each {@link IndexProperties#toKey(int) index keyFunction} has an id,
+ * and this id matches the id of the indexer;
+ * each keyFunction in {@link IndexProperties} is associated with a single indexer.
+ * <p>
+ * Comparison joiners result in a single indexer each,
+ * whereas equal joiners will be merged into a single indexer if they are consecutive.
+ * In the latter case,
+ * a composite keyFunction is created of type {@link Pair}, {@link TriTuple},
+ * {@link Quadruple} or {@link IndexerKey},
+ * based on the length of the composite keyFunction (number of equals joiners in sequence).
+ *
+ * <ul>
+ * <li>Example 2: For an EQUAL+LESS_THAN joiner,
+ * there are two indexers in the chain with keyFunction length of 1 each.</li>
+ * <li>Example 3: For an LESS_THAN+EQUAL+EQUAL joiner,
+ * there are still two indexers,
+ * but the second indexer's keyFunction length is 2.</li>
+ * <li>Example 4: For an LESS_THAN+EQUAL+EQUAL+LESS_THAN joiner,
+ * there are three indexers in the chain,
+ * and the middle one's keyFunction length is 2.</li>
+ * </ul>
+ *
+ * @param <Right_>
+ */
+public class IndexerFactory<Right_> {
 
-    private final JoinerType[] joinerTypes;
+    private final AbstractJoiner<Right_> joiner;
+    private final NavigableMap<Integer, JoinerType> joinerTypeMap;
 
-    public IndexerFactory(AbstractJoiner joiner) {
-        int joinerCount = joiner.getJoinerCount();
-        joinerTypes = new JoinerType[joinerCount];
-        for (int i = 0; i < joinerCount; i++) {
-            JoinerType joinerType = joiner.getJoinerType(i);
-            switch (joinerType) {
-                case EQUAL:
-                case LESS_THAN:
-                case LESS_THAN_OR_EQUAL:
-                case GREATER_THAN:
-                case GREATER_THAN_OR_EQUAL:
-                    break;
-                default:
-                    throw new UnsupportedOperationException("Unsupported joiner type (" + joinerType + ").");
+    public IndexerFactory(AbstractJoiner<Right_> joiner) {
+        this.joiner = joiner;
+        var joinerCount = joiner.getJoinerCount();
+        if (joinerCount < 2) {
+            joinerTypeMap = null;
+        } else {
+            joinerTypeMap = new TreeMap<>();
+            for (var i = 1; i <= joinerCount; i++) {
+                var joinerType = i < joinerCount ? joiner.getJoinerType(i) : null;
+                var previousJoinerType = joiner.getJoinerType(i - 1);
+                if (joinerType != JoinerType.EQUAL || previousJoinerType != joinerType) {
+                    // Equal joiner may build composite keyFunction with preceding equal joiner(s).
```

Comment:
The condition and the comment made me think there is a bug at the first glance; after carefully looking in the code I understand we intentionally skip the map for Equal joiners so that the index increases through the iteration.
Maybe a bit more explanatory comment would help.